core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationOutputProcessor.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.aggregation.partial;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Ints;
+import io.trino.operator.aggregation.AggregatorFactory;
+import io.trino.operator.aggregation.GroupedAggregator;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.spi.type.Type;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Creates final output of the partial aggregation step based on either aggregated or raw input.
+ * <p>
+ * Partial aggregation step output has additional channels used to send
+ * raw input data when the partial aggregation is disabled.
+ * Those channels need to be present in the output, albeit empty, even if the
+ * partial aggregation is enabled.
+ * <p>
+ * The additional channels are:
+ * - mask channels used by any aggregation
+ * - every aggregation input channels that are not already in hash channels
+ * - a boolean rawInputMask channel that contains decision which input, aggregated or raw, should be used for a given position.
+ */
+public class PartialAggregationOutputProcessor
+{
+    private final List<GroupedAggregator> groupedAggregators;
+    private final List<Type> aggregationInputTypes;
+    private final int[] hashChannels;
+    private final int[] aggregationInputChannels;
+    private final int[] maskBlockChannels;
+    private final int finalBlockCount;
+
+    public static PartialAggregationOutputProcessor create(List<Integer> groupByChannels, Optional<Integer> hashChannel, List<AggregatorFactory> aggregatorFactories, List<? extends Type> aggregationInputTypes, List<Integer> aggregationInputChannels)
```

Comment:
`create` method is not needed. Just use `PartialAggregationOutputProcessor` constructor